src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+            {
+                CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+                return Binder.LeastSpecificType(_conversions, candidates, ref useSiteInfo);
+            }
+
+            public override BoundNode? VisitNegatedPattern(BoundNegatedPattern node)
+            {
+                var savedNegated = _negated;
+                _negated = !_negated;
+                var result = this.Visit(node.Negated);
+                _negated = savedNegated;
+                return result;
+            }
+
+            public BoundPattern NegateIfNeeded(BoundPattern node)
+            {
+                if (!_negated)
+                {
+                    return node;
+                }
+
+                if (node is BoundNegatedPattern { Negated: var negated })
+                {
+                    return negated;
+                }
+
+                var result = new BoundNegatedPattern(node.Syntax, node, node.InputType, narrowedType: node.InputType);
+                if (node.WasCompilerGenerated)
+                {
+                    result.MakeCompilerGenerated();
+                }
+
+                return result;
+            }
+
+            public override BoundNode? VisitTypePattern(BoundTypePattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            public override BoundNode? VisitConstantPattern(BoundConstantPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            public override BoundNode? VisitDiscardPattern(BoundDiscardPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            public override BoundNode? VisitRelationalPattern(BoundRelationalPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            private BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern)
+            {
+                return WithInputTypeCheckIfNeeded(pattern, _inputType);
+            }
+
+            private static BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern, TypeSymbol inputType)
+            {
+                // Produce `PatternInputType and pattern` given a new input type
+
+                if (pattern.InputType.Equals(inputType, TypeCompareKind.AllIgnoreOptions))
+                {
+                    return pattern;
+                }
+
+                BoundPattern typePattern = new BoundTypePattern(pattern.Syntax,
+                    new BoundTypeExpression(pattern.Syntax, aliasOpt: null, pattern.InputType),
+                    isExplicitNotNullTest: false, inputType, narrowedType: pattern.InputType);
+
+                var result = new BoundBinaryPattern(pattern.Syntax, disjunction: false, left: typePattern, right: pattern, inputType, pattern.NarrowedType);
```

Comment:
>BoundBinaryPattern [](http://example.com/codeflow?start=33&length=18)

It feels like if the input is a type pattern, we could simply adjust its input type.

Reply:
It looks like recursive patterns can be updated too. Here is what I have in mind:
```
                if (pattern is BoundTypePattern typePattern1)
                {
                    return typePattern1.Update(typePattern1.DeclaredType, typePattern1.IsExplicitNotNullTest, inputType, typePattern1.NarrowedType);
                }

                if (pattern is BoundRecursivePattern recursivePattern)
                {
                    return recursivePattern.Update(
                        recursivePattern.DeclaredType ??
                            new BoundTypeExpression(recursivePattern.Syntax, aliasOpt: null, recursivePattern.InputType.StrippedType()),
                        recursivePattern.DeconstructMethod, recursivePattern.Deconstruction,
                        recursivePattern.Properties, recursivePattern.IsExplicitNotNullTest,
                        recursivePattern.Variable, recursivePattern.VariableAccess,
                        inputType, recursivePattern.NarrowedType);
                }
```

Reply:
Discard as well:
```
                if (pattern is BoundDiscardPattern discardPattern)
                {
                    return discardPattern.Update(inputType, inputType);
                }
```

Reply:
Given the above, it might be worth handling negated pattern specially too:
```
                if (pattern is BoundNegatedPattern negatedPattern)
                {
                    return negatedPattern.Update(
                        WithInputTypeCheckIfNeeded(negatedPattern.Negated, inputType),
                        inputType, inputType);
                }
```

Reply:
For constant and relational patterns would it be valid to simply adjust `InputType` in the node without creating an `and`?

Reply:
After adding the code that digs through BoundNegatedPattern, I noticed that we can get here when `pattern` is a binary pattern. It feels like this will be acceptable only when we are readjusting previously adjusted input type. Therefore, I think adding the following code would make sense:
```
                if (pattern is BoundBinaryPattern { Disjunction: false, Left: BoundTypePattern typePattern2 } binary)
                {
                    // We expect to run into only into 'and' that we earlier created below to adjust an input type
                    // Now we are adjusting it again, likely undoing the adjustment
                    Debug.Assert(typePattern2 is { WasCompilerGenerated: true, IsExplicitNotNullTest: false });
                    Debug.Assert(binary.Right is not (BoundBinaryPattern or BoundNegatedPattern));

                    var adjustedRight = WithInputTypeCheckIfNeeded(binary.Right, inputType);
                    Debug.Assert(adjustedRight == binary.Right); // Do we have a test covering a scenario that fails this assert?
                    return adjustedRight;
                }

                Debug.Assert(pattern is not BoundBinaryPattern);
```

The situation is when this is happening is first we are adjusting input type when we rewrite a sub-pattern, and the adjusting the type again when we are pulling `and`s and `or`s out of the sub-pattern. I think we should seriously look for a way of unifying both processes and adjust input type only once. The first adjustment might be not under `not` and the process of pulling `and`s and `or`s out of the sub-pattern, simply going to split the `and` that we are creating here. resulting in an unnecessary type sub-pattern. After the process is unified, I do not expect us run into BoundBinaryPattern in this function at all.